utils/near-rate-limiter/src/framed_read.rs

```diff
@@ -0,0 +1,380 @@
+impl<T, D> Stream for ThrottledFrameRead<T, D>
+where
+    T: AsyncRead,
+    D: Decoder,
+{
+    type Item = Result<D::Item, D::Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        self.project().inner.poll_next(cx)
+    }
+}
+
+pub(crate) struct ReadFrame {
+    pub(crate) eof: bool,
+    pub(crate) is_readable: bool,
+    pub(crate) buffer: BytesMut,
+}
+
+impl Default for ReadFrame {
+    fn default() -> Self {
+        Self { eof: false, is_readable: false, buffer: BytesMut::with_capacity(INITIAL_CAPACITY) }
+    }
+}
+
+impl<T, U, R> Stream for FramedImpl<T, U, R>
+where
+    T: AsyncRead,
+    U: Decoder,
+    R: BorrowMut<ReadFrame>,
+{
+    type Item = Result<U::Item, U::Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut pinned = self.project();
+        let state: &mut ReadFrame = pinned.state.borrow_mut();
+        loop {
+            while let task::Poll::Ready(_) = pinned.receiver.poll_recv(cx) {
+                // pop all elements from queue, to prevent memory leak
+            }
+            while !pinned.rate_limiter.is_ready() {
+                // This will cause us to subscribe to notifier when something gets pushed to
+                // `pinned.receiver`. If there is an element in the queue, we will check again.
+                ready!(pinned.receiver.poll_recv(cx));
+            }
+
+            // Repeatedly call `decode` or `decode_eof` as long as it is
+            // "readable". Readable is defined as not having returned `None`. If
+            // the upstream has returned EOF, and the decoder is no longer
+            // readable, it can be assumed that the decoder will never become
+            // readable again, at which point the stream is terminated.
+            if state.is_readable {
+                if state.eof {
+                    let frame = pinned.codec.decode_eof(&mut state.buffer)?;
+                    return Poll::Ready(frame.map(Ok));
+                }
+                trace!("attempting to decode a frame");
+
+                if let Some(frame) = pinned.codec.decode(&mut state.buffer)? {
+                    trace!("frame decoded from buffer");
+                    return Poll::Ready(Some(Ok(frame)));
+                }
+
+                state.is_readable = false;
+            }
+
+            assert!(!state.eof);
+
+            // Otherwise, try to read more data and try again. Make sure we've
+            // got room for at least one byte to read to ensure that we don't
+            // get a spurious 0 that looks like EOF
+            state.buffer.reserve(1);
+
+            let bytect = match poll_read_buf(pinned.inner.as_mut(), cx, &mut state.buffer)? {
+                Poll::Ready(ct) => ct,
+                // We are waiting for more data on TCP socket.
+                Poll::Pending => {
+                    return Poll::Pending;
+                }
+            };
```

Comment:
Let's use the `ready!` macro here. Explicit `return Pending` makes the reader think about potential lost wakeup (returning Pending without arranging for the waker to be notified). With `ready!`, this problem is impossible by construction.

Reply:
That's a good idea. I implemented this the same way as `FrameReader` is implemented in `tokio_utils`. Maybe we should update it there as well.